gdal/port/cpl_json.h

```diff
@@ -0,0 +1,196 @@
+ * and/or sell copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included
+ * in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+ * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+ * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ ****************************************************************************/
+
+#ifndef CPL_JSON_H_INCLUDED
+#define CPL_JSON_H_INCLUDED
+
+#include "cpl_string.h"
+#include "cpl_progress.h"
+
+/**
+ * \file cpl_json.h
+ *
+ * Interface for read and write JSON documents
+ */
+
+/*! @cond Doxygen_Suppress */
+typedef void *JSONObjectH;
+
+CPL_C_START
+
+class CPLJSONArray;
+/*! @endcond */
+
+/**
+ * @brief The CPLJSONArray class holds JSON object from CPLJSONDocument
+ */
+class CPL_DLL CPLJSONObject
+{
+    friend class CPLJSONArray;
+    friend class CPLJSONDocument;
+public:
+    /**
+     * Json object types
+     */
+    enum Type {
+        Null,
+        Object,
+        Array,
+        Boolean,
+        String,
+        Integer,
+        Long,
+        Double
+    };
+
+public:
+/*! @cond Doxygen_Suppress */
+    CPLJSONObject();
+    explicit CPLJSONObject(const char *pszName, const CPLJSONObject &oParent);
+    ~CPLJSONObject();
+    CPLJSONObject(const CPLJSONObject &other);
+    CPLJSONObject &operator=(const CPLJSONObject &other);
+
+#if !_MSC_VER
+private:
+#endif // ! _MSC_VER
+    explicit CPLJSONObject(const CPLString &soName, JSONObjectH poJsonObject);
+/*! @endcond */
+
+public:
+    // setters
+    void Add(const char *pszName, const CPLString &soValue);
+    void Add(const char *pszName, const char *pszValue);
+    void Add(const char *pszName, double dfValue);
+    void Add(const char *pszName, int nValue);
+    void Add(const char *pszName, GInt64 nValue);
+    void Add(const char *pszName, const CPLJSONArray &oValue);
+    void Add(const char *pszName, const CPLJSONObject &oValue);
+    void Add(const char *pszName, bool bValue);
+
+    void Set(const char *pszName, const char *pszValue);
+    void Set(const char *pszName, double dfValue);
+    void Set(const char *pszName, int nValue);
+    void Set(const char *pszName, GInt64 nValue);
+    void Set(const char *pszName, bool bValue);
+
+/*! @cond Doxygen_Suppress */
+    JSONObjectH GetInternalHandle() const { return m_poJsonObject; }
+/*! @endcond */
+
+    // getters
+    const char *GetString(const char *pszName, const char *pszDefault) const;
```

Comment:
Wouldn't default values for those default arguments would make sense ?

Reply:
I'm afraid what compiler will decide this two methods are the same: 
const char *GetString(const char *pszName, const char *pszDefault = "") const;
const char *GetString(const char *pszDefault = "") const;

Reply:
Good point. Perhaps just on the methods with a single argument then ?

Reply:
Hum or have  different names, which actually better reflect what the method do:
```
const char* GetString(const char* pszName, const char* pszDefault = nullptr)
const char* ToString(const char* pszDefault = nullptr)
```

Reply:
Great idea! I like it.

Reply:
Done